### PR TITLE
Using ZipArchiveLoader from torchdata

### PR DESCRIPTION
**Pull request by fabhertz95**
More than a PR, this is to show how to use torchdata archive utilities and to discuss next steps. I would like to add some other things before merging it. 

I used `ZipArchiveLoader`, which allow to cycle through the files inside the ZIP archive. Does it make any sense to you, @biphasic? See [this line from torchvision](https://github.com/pytorch/vision/blob/528651a031a08f9f97cc75bd619a326387708219/torchvision/prototype/datasets/_builtin/mnist.py#L98) and [this dataset class](https://github.com/pytorch/vision/blob/main/torchvision/prototype/datasets/utils/_dataset.py).

Edit: I tried to implement an STMNIST class inspired by torchvision new datapipe implementation. We can talk about it if you want, it's pretty complicated. 

This is an usage example.
```python
from torchdata.datapipes.iter import (
    ZipArchiveLoader,
    UnZipper,
    FileLister, 
    FileOpener,
    Filter
)
from scipy.io import loadmat

filepath = "./STMNIST.zip"

def is_mat(x):
    return x[0].endswith(".mat") and "LUT" not in x[0]

dp = FileLister(filepath)
dp = FileOpener(dp, mode="b")
dp = ZipArchiveLoader(dp).filter(is_mat)
dp_fpath, dp_data = UnZipper(dp, sequence_length=2)
iter_dp = iter(dp)
fpath, stream = next(iter_dp)
fpath, stream

```

    ('STMNIST.zip/data_submission/9/9_ch0_622_spiketrain.mat',
     StreamWrapper<<zipfile.ZipExtFile name='data_submission/9/9_ch0_622_spiketrain.mat' mode='r' compress_type=deflate>>)

```python
mat = loadmat(stream)
a = mat['spiketrain']
a
```

    array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
            0.        ],
           [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
            0.        ],
           [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
            0.        ],
           ...,
           [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
            0.        ],
           [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
            0.        ],
           [0.07939189, 0.0793919 , 0.07939192, ..., 1.99187338, 1.99187339,
            1.99187341]])

### CODE REVIEW

**Comment by fabhertz95** [comment]
Well, actually there seems to be no slowdown if the decoding function is included in the pipe... If you remove the encoding step, the ratio zipped/unzipped load times is ~6.

```python
from torchdata.datapipes.iter import (
    ZipArchiveLoader,
    UnZipper,
    FileLister, 
    FileOpener,
    Filter
)
from scipy.io import loadmat
import numpy as np
import timeit

def is_mat_tuple(x):
    return x[0].endswith(".mat") and "LUT" not in x[0]

def is_mat(x):
    return x.endswith(".mat") and "LUT" not in x

def spiketrain_to_array(matfile):
    # Transposing since the order is (address, event),
    # but we like (event, address).
    mat = loadmat(matfile)
    spiketrain = mat["spiketrain"].T
    # Separating coordinates and timestamps.
    spikes, timestamps = spiketrain[:, :-1], spiketrain[:, -1]
    # Getting events addresses.
    # First entry -> Event number.
    # Second entry -> Event address in [0,100).
    events_nums, events_addrs = spikes.nonzero()
    # Mapping addresses to 2D coordinates.
    # The mapping is (x%address, y//address), from the paper.
    events = np.zeros((len(events_nums)), dtype=dtype)
    events["x"] = events_addrs % sensor_size[0]
    events["y"] = events_addrs // sensor_size[1]
    # Converting floating point seconds to integer microseconds.
    events["t"] = (timestamps[events_nums] * 1e6).astype(int)
    # Converting -1 polarities to 0.
    events["p"] = np.max(spikes[(events_nums, events_addrs)], 0).astype(int)
    return events

###########################################################################

zip_path = "./STMNIST.zip"
datapath = "./data_submission"
sensor_size = (10, 10, 2)
dtype = np.dtype([("x", int), ("y", int), ("t", int), ("p", int)])

# Zipped.
zip_dp = FileLister(zip_path)
zip_dp = FileOpener(zip_dp, mode="b")
zip_dp = ZipArchiveLoader(zip_dp).filter(is_mat_tuple)
_, zip_dp = UnZipper(zip_dp, sequence_length=2)
zip_dp = zip_dp.map(spiketrain_to_array)
iter_zip_dp = iter(zip_dp)

# Unzipped.
dp = FileLister(datapath, recursive=True).filter(is_mat)
dp = dp.map(spiketrain_to_array)
iter_dp = iter(dp)

fn = lambda: next(iter_dp)
zip_fn = lambda: next(iter_zip_dp)

avg_time = timeit.Timer(fn).timeit(number=1000)
print(f"Average load time for simple datapipe: {avg_time:.3f} ms")
zip_avg_time = timeit.Timer(zip_fn).timeit(number=1000)
print(f"Average load time for zipped datapipe: {zip_avg_time:.3f} ms")
print(f"Ratio between load times (ZIP v.s. unzipped): {zip_avg_time/avg_time:.2f}x.")
```

    Average load time for simple datapipe: 3.624 ms
    Average load time for zipped datapipe: 3.428 ms
    Ratio between load times (ZIP v.s. unzipped): 0.95x.

File: tonic/prototype/datasets/stmnist.py [hunk 1]
```diff
@@ -1,113 +1,111 @@
-"""
-Novel neuromorphic Spiking Tactile MNIST (ST-MNIST) dataset, which comprises handwritten digits obtained by human participants writing on a neuromorphic tactile sensor array.
-The original paper can be found at https://arxiv.org/abs/2005.04319.
-Data is provided with the MAT format. 
-Download of the compressed dataset has to be done by the user by accessing https://scholarbank.nus.edu.sg/bitstream/10635/168106/2/STMNIST%20dataset%20NUS%20Tee%20Research%20Group.zip, where a form has to be compiled. Then, the path to the ZIP archive has to be provided to the stmnist() function root argument.
-"""
```
Comment (biphasic) [inline tonic/prototype/datasets/stmnist.py:6]: can we keep this comment? I think this is helpful

File: tonic/prototype/datasets/stmnist.py [hunk 2]
```diff
@@ -1,113 +1,111 @@
...
+        events["y"] = events_addrs // self.sensor_size[1]
+        # Converting floating point seconds to integer microseconds.
+        events["t"] = (timestamps[events_nums] * 1e6).astype(int)
+        # Converting -1 polarities to 0.
+        events["p"] = np.max(spikes[(events_nums, events_addrs)], 0).astype(int)
+        return events
+
+
+class STMNIST(Dataset):
+
```
Comment (biphasic) [inline tonic/prototype/datasets/stmnist.py:69]: maybe we can have the docstring here

**Review by biphasic — CHANGES_REQUESTED** [review]
This looks good! Please add back the docstring that explains the dataset and parameters (unless I missed it to be somewhere else). Thank you!

**Review by biphasic — APPROVED** [review]
looking good!

**Comment by biphasic** [comment]
should I merge?

**Comment by fabhertz95** [comment]
I just have one doubt. I benchmarked over 1000 samples the read speed. 
image
Right now, we are using the archive loaders from torchdata to iterate through the dataset. With respect to the original Tonic datasets, hence, we perform no decompression. In your opinion, should we provide the option to decompress the archive and use the decompressed file? For NMNIST there is a 5x speedup. 

At the same time, once the tensors are cached to the GPU or somewhere else, who cares? :laughing: I wanted to know your opinion about this, to see if it was worth to spend time adding a `decompress` argument to the dataset class.

**Comment by biphasic** [comment]
oh I see, so for some datasets we still needs to decompress the data. I think that makes sense and a 5x speedup is definitely worth a little bit of extra disk space. You're right that once tensors are cached it doesn't matter so much, but it still matters for the first epoch (during caching) and for people who work directly with the events instead of tensors (and do not use caching at all). I think that the default setting for decompression should be True for datasets like NMNIST, CIFAR10DVS, DVS Gesture and the like because they're not that huge at the end of the day and fast loading is what we want to achieve! 

Recently I re-worked the DSEC dataset a bit and it's a quite large one. So to save some disk space, I just delete the zip file after extraction by default, see here: https://github.com/neuromorphs/tonic/blob/develop/tonic/datasets/dsec.py#L344
So my answer is: If it's (quite a bit) faster to load the decompressed files, we should do that. If the dataset is huge, we can think of deleting zips after extraction or providing a decompression=False flag

**Comment by fabhertz95** [comment]
I think you are absolutely right. Then, let's decompress the data by default. I want to understand if this can be accomplished trough torchdata utilities, so that we have to maintain only Tonic-related code and we can rely on torchdata for the most basic tasks. There must be a way since all the datasets in `torchvision.prototype` do not decompress the data in advance but use the archive loaders from torchdata. I'll investigate a little bit :laughing:

**Comment by biphasic** [comment]
thanks a lot @fabhertz95 !

**Comment by fabhertz95** [comment]
I managed to use datapipes for the decompression (no code to maintain for us, yeeee :laughing:) but between the old and new dataset, on NMNIST, there is a 30% slowdown in the read performance. For STMNIST, for instance, it makes no difference to keep the directory compressed or not in terms of speed.

**Comment by biphasic** [comment]
@fabhertz95 this starts to look excellent 👌 the readme is great and I'm pleased to read that we can make use of lots of torchdata code. Can you explain to me why the new version of nmnist is still slower? Is it that torchdata's decompression dp is releasing files iteratively and therefore slowing things down? Apologies as I'm only reading all this on my phone these days

**Comment by fabhertz95** [comment]
I onestly do not know :disappointed: If `keep_compressed=False`, the dataset in uncompressed in a dedicated folder at object creation, in the following way: 

```python 
def _uncompress(self, dp):
        dp = FileLister(path_to_zip)
        dp = FileOpener(dp, mode="rb")
        # Iterating through the ZIP archive.
        dp = ZipArchiveLoader(dp)
        # Saving the file to disk. This code is not completely correct, but it lets you understand what happens.
        dp = Saver(dp, filepath_fn=fn_that_provides_path_to_destination_folder)
        # Then we iterate through the whole dataset to performs uncompression and saving with this datapipe. 
        for x in dp:
            pass
        # Now we return to the class the datapipe that iterates through the uncompressed files.
        dp = FileLister(path_to_uncompressed_folder)
        dp = FileOpener(dp, mode="rb")
        return dp
```
Hence, the dataset receives a datapipe that iterates through normal files in a folder. Using the profiler, it turns out that the problem is the overhead introduced by the datapipe. Like in `FileLister`, which performs substantially a `for` loop with `os.walk`, which is implemented directly in the old dataset. The advantage with datapipes, at this point, seems to be that we have to write much less code (no unzipping routine, file listing and so on) and we do not have to maintain code not related to Tonic itself.

**Comment by biphasic** [comment]
A performance hit of 11% is worth it. This is great stuff @fabhertz95 . Should I merge?

**Comment by fabhertz95** [comment]
I think the code is decent now. Let's merge like there is no tomorrow :laughing:

**Comment by biphasic** [comment]
One more thing, is there a reason we're not using this anymore https://github.com/neuromorphs/tonic/blob/beef9808639693762893b9f17106a6d9ea61bc74/tonic/io.py#L157

**Comment by biphasic** [comment]
Like in the old version

**Comment by fabhertz95** [comment]
> One more thing, is there a reason we're not using this anymore
> 
> https://github.com/neuromorphs/tonic/blob/beef9808639693762893b9f17106a6d9ea61bc74/tonic/io.py#L157

Yes, the reason is that `read_mnist_file` wants the file in input, while the datapipe provides a binary stream. In fact, if you look here:

https://github.com/fabhertz95/tonic/blob/2bc344152088a5153a3db54167aa35a403bc1f6e/tonic/prototype/datasets/nmnist.py#L47

you can see that I read the already opened binary file using `np.frombuffer()`. 

In the latest commit, with uncompressed archives I use `read_mnist_file`, by removing the `FileOpener` datapipe and propagating `keep_compressed` to the `NMNISTFileReader` class. Does it sound good?

**Comment by biphasic** [comment]
Sounds good 👍 merging now like there is no tomorrow

**Pull request merged**